src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Helpers.cs

```diff
@@ -0,0 +1,145 @@
+                return certificates;
+            }
+        }
+
+        /// <summary>
+        /// Note: Due to its use of X509Certificate2Collection.Find(), this returns a cloned certificate. So the caller should dispose it.
+        /// </summary>
+        public static X509Certificate2 TryFindMatchingCertificate(this X509Certificate2Collection certs, SubjectIdentifier recipientIdentifier)
+        {
+            SubjectIdentifierType recipientIdentifierType = recipientIdentifier.Type;
+            switch (recipientIdentifierType)
+            {
+                case SubjectIdentifierType.IssuerAndSerialNumber:
+                    {
+                        X509IssuerSerial issuerSerial = (X509IssuerSerial)(recipientIdentifier.Value);
+
+                        using (ClonedCertificates matches1 = certs.FindCertificates(X509FindType.FindBySerialNumber, issuerSerial.SerialNumber))
+                        {
+                            using (ClonedCertificates matches2 = matches1.FindCertificates(X509FindType.FindByIssuerDistinguishedName, issuerSerial.IssuerName))
+                            {
+                                // Desktop compat: We do not complain about multiple matches. Just take the first one and ignore the rest.
+                                return matches2.Any ? matches2.ClaimResult() : null;
+                            }
+                        }
+                    }
+
+                case SubjectIdentifierType.SubjectKeyIdentifier:
+                    {
+                        string ski = (string)(recipientIdentifier.Value);
+                        using (ClonedCertificates matches = certs.FindCertificates(X509FindType.FindBySubjectKeyIdentifier, ski))
+                        {
+                            // Desktop compat: We do not complain about multiple matches. Just take the first one and ignore the rest.
+                            return matches.Any ? matches.ClaimResult() : null;
+                        }
+                    }
+
+                default:
+                    // RecipientInfo's can only be created by this package so if this an invalid type, it's the package's fault.
+                    Debug.Fail($"Invalid recipientIdentifier type: {recipientIdentifierType}");
+                    throw new CryptographicException();
+            }
+        }
+
+        /// <summary>
+        /// Useful helper for "upgrading" well-known CMS attributes to type-specific objects such as Pkcs9DocumentName, Pkcs9DocumentDescription, etc. 
+        /// </summary>
+        public static Pkcs9AttributeObject CreateBestPkcs9AttributeObjectAvailable(Oid oid, byte[] encodedAttribute)
+        {
+            Pkcs9AttributeObject attributeObject = new Pkcs9AttributeObject(oid, encodedAttribute);
+            switch (oid.Value)
+            {
+                case Oids.DocumentName:
+                    attributeObject = Upgrade<Pkcs9DocumentName>(attributeObject);
```

Comment:
Rather than create, upgrade (which is create, create, copy, return, GC), can this just be return new Pkcs9DocumentName(encodedAttribute), repeat for the other special types, and default to the Pkcs9AttributeObject ctor?

In spot checking it didn't look like they do anything special in CopyFrom.